rotkehlchen/exchanges/kraken.py

```diff
@@ -363,27 +409,79 @@ def _query_private(self, method: str, req: dict | None = None) -> dict | str:
             message=message,
             digest_algorithm=hashlib.sha512,
         )
-        self.session.headers.update({
-            'API-Sign': signature,
-        })
         try:
             response = self.session.post(
                 KRAKEN_BASE_URL + urlpath,
                 data=post_data.encode(),
                 timeout=CachedSettings().get_timeout_tuple(),
+                headers={'APIKey': self.api_key, 'API-Sign': signature},
             )
         except requests.exceptions.RequestException as e:
             raise RemoteError(f'Kraken API request failed due to {e!s}') from e
         self._manage_call_counter(method)
 
-        return _check_and_get_response(response, method)
+        decoded_json = _check_and_get_response(response, method)
+
+        if isinstance(decoded_json, str):
+            return decoded_json
+
+        result = decoded_json.get('result', None)
+        if result is None:
+            if method == 'Balance':
+                return {}
+
+            raise RemoteError(f'Missing result in kraken response for {method}')
+
+        return result
 
     # ---- General exchanges interface ----
     @protect_with_lock()
     @cache_response_timewise()
     def query_balances(self) -> ExchangeQueryBalances:
+        returned_balances: dict[AssetWithOracles, Balance] = {}
+        spot_balances, spot_msg = self.query_spot_balances()
+        if not self._has_futures_keys():
+            log.debug('Kraken Futures keys are not set, returning only spot balances')
+            return spot_balances, spot_msg
+        else:
+            futures_balances, futures_msg = self.query_futures_balances()
+            if spot_balances is None:
```

Comment:
this can be improved. first handle the case where both are present and return. then other cases can follow.